plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -139,6 +139,51 @@ nonZeroArg :: (Integer -> Integer -> Integer) -> Integer -> Integer -> Evaluatio
 nonZeroArg _ _ 0 = EvaluationFailure
 nonZeroArg f x y = EvaluationSuccess $ f x y
 
+{- Note [Constants vs built-in functions]
+A constant is anything of a built-in type. For example, 'Integer' is a built-in type, so anything of
+type 'Integer' is a constant.
+
+On the contrary a built-in function can't be of a built-in type, because the type of a built-in
+function is always of either the @all a. b@ form or the @a -> b@ one, none of which is a built-in
+type. This is checked by the machinery, so if the user tries to add a built-in function that is not
+of one of these forms, they'll get a nice custom type error.
+
+A built-in function is associated with its Haskell implementation: there can be many built-in
+functions of the same type, all doing different things, and there can be infinitely more _definable_
+built-in functions of the same type that are not built-in functions nonetheless, because we didn't
+register them as such by providing a Haskell implementation for each of them. This is the difference
+between constants and built-in functions: the set of constants (infinite in our case) depends solely
+on the set of available built-in types (also infinite in our case, because we have @Integer@,
+@[Integer]@, @[[Integer]]@ etc), while the set of built-in functions is defined by explicitly
+assigning each member a specific name and an associated with it Haskell implementation. It is
+theoretically possible to have an infinite set of built-in functions, but we neither do that nor
+need it, hence our set of built-in functions is finite.
+
+The rule of thumb is: constants are raw data and built-in functions are, well, functions.
+
+@(:)@ works as follows: it takes two constants wrapped as values, extracts an integer from the first
```

Comment:
Can you confirm/clarify (if it hasn't been done elsewhere) that builtin functions must take constants as arguments, and not arbitrary terms like a function?

Reply:
No, it's `(:)` that only takes constants. `ifThenElse` (being a builtin) takes arbitrary terms as its second and third arguments. There's a lot of docs below this note elaborating on the whole alignment. Just amended a part of another Note to become

> The idea is simple: in order to apply the denotation of a builtin expecting, say,
> an 'Integer' constant we need to actually extract that 'Integer' from the AST of the given value,
> but if the denotation is polymorphic over the type of its argument, then we don't need to extract
> anything, we can just pass the AST of the value directly to the denotation (which means the value
> doesn't have to be a 'Constant', it can be completely arbitrary).

Thanks for the suggestion.